Technical Edit of Baldy Release (#1570)
cntt-n/CNTT: doc/ref_arch/kubernetes/chapters/chapter03.md

<a name="3.1"></a>
## 3.1 Introduction

The CNTT Kubernetes Reference Architecture (RA) is intended to be an industry standard independent Kubernetes reference architecture that is not tied to any specific offering or distribution. No vendor-specific enhancements are required in order to achieve conformance to CNTT principles; conformance is achieved by using upstream components or features that are developed by the open source community. This allows operators to have a common Kubernetes-based architecture that supports any conformant VNF or CNF deployed on it to operate as expected. The purpose of this chapter is to outline all the components required to provide Kubernetes in a consistent and reliable way.  The specification of how to use these components is detailed in [Chapter 04](chapter04.md).

> Kubernetes is already a well documented and widely deployed Open Source project managed by The Cloud Native Computing FOundation (CNCF). Full documentation of the code and project can be found at [https://kubernetes.io/docs/home/](https://kubernetes.io/docs/home/). The  the following chapters will only describe the specific features required by the CNTT Reference Architecture, and how they would be expected to be implemented. For any information related to standard Kubernates features and capabilities, refer back to the standard Kubernates documentation.

Reviewer: ```suggestion
Kubernetes is already a well documented and widely deployed Open Source project managed by the Cloud Native Computing Foundation (CNCF). Full documentation of the Kubernetes code and project can be found at [https://kubernetes.io/docs/home/](https://kubernetes.io/docs/home/). The following chapters will only describe the specific features required by the CNTT Reference Architecture, and how they would be expected to be implemented. For any information related to standard Kubernates features and capabilities, refer back to the standard Kubernates documentation.
```

<a name="3.1"></a>
## 3.1 Introduction

> The CNTT Kubernetes Reference Architecture (RA) is intended to be an industry standard independent Kubernetes reference architecture that is not tied to any specific offering or distribution. No vendor-specific enhancements are required in order to achieve conformance to CNTT principles; conformance is achieved by using upstream components or features that are developed by the open source community. This allows operators to have a common Kubernetes-based architecture that supports any conformant VNF or CNF to be deployed and operated as expected. The purpose of this chapter is to outline all the components required to provide Kubernetes in a consistent and reliable way.  The specification of how to use these components is detailed in [Chapter 04](chapter04.md).

Reviewer: ```suggestion
The CNTT Kubernetes Reference Architecture (RA) is independent of any specific offering or distribution. No vendor-specific enhancements are required in order to achieve conformance to CNTT principles; conformance is achieved by using upstream components or features that are developed by the open source community. This allows operators to have a common Kubernetes-based architecture that supports any conformant VNF or CNF to be deployed and operate as expected. The purpose of this chapter is to outline all the components required to provide Kubernetes in a consistent and reliable way.  The specification of how to use these components is detailed in [Chapter 04](chapter04.md).
```

- Kernel space
- User space

> The Kernel is the tightly controlled space that provides an API via system calls to applications running in the user space (which usually have their own southbound interface in an interpreter or libraries). Key containerisation capabilities such as Control Groups (cgroups) and namespaces are kernel features, and are used and managed by the container runtime in order to provide isolation between the user space processes, which would also include the container itself as well as any processes running within it. The security of the Host OS and its relationship to the container and the applications running within the container or containers is essential to the overall security posture of the entire system, and must be appropriately secured to ensure processes running in one container cannot escalate their privileges or otherwise affect processes running in an adjacent container.  An example and more details of this concept can be found in [chapter 6](./chapter06.md)).

Reviewer: ```suggestion
The Kernel is the tightly controlled space that provides an API to applications running in the user space (which usually have their own southbound interface in an interpreter or libraries). Key containerisation capabilities such as Control Groups (cgroups) and namespaces are kernel features, and are used and managed by the container runtime in order to provide isolation between the user space processes, which would also include the container itself as well as any processes running within it. The security of the Host OS and its relationship to the container and the applications running within the container or containers is essential to the overall security posture of the entire system, and must be appropriately secured to ensure processes running in one container cannot escalate their privileges or otherwise affect processes running in an adjacent container.  An example and more details of this concept can be found in [chapter 6](./chapter06.md)).
```

> It is important to note that the container runtime itself is also a set of processes that run in user space, and therefore also interact with the kernel via system calls.  Many diagrams will show containers as running on top of the runtime, or inside the runtime.  More accurately, the containers themselves are simply processes running within an OS, the container runtime is simply another set of processes that are used to manage those containers (pull, run, delete, etc.), and the kernel features required to provide the isolation (cgroups, namespaces, filesystems, etc.) mechanisms between the components.

Reviewer: ```suggestion
It is important to note that the container runtime itself is also a set of processes that run in user space, and therefore also interact with the kernel via system calls.  Many diagrams will show containers as running on top of the runtime, or inside the runtime.  More accurately, the containers themselves are simply processes running within an OS, the container runtime is simply another set of processes that are used to manage these containers (pull, run, delete, etc.), and the kernel features required to provide the isolation mechanisms  (cgroups, namespaces, filesystems, etc.) between the components.
```

<a name="3.2.1.1"></a>
#### 3.2.1.1 Memory management

> Relate back to features described in the RM [here](../../../ref_model/chapters/chapter05.md#521-virtual-compute). Note that the RM appears to be missing Memory-based HW profile features [here](../../../ref_model/chapters/chapter05.md#54-nfvi-hw-profiles-features-and-requirements).

> The Reference Model requires the support of Huge Pages in `i.cap.018` which is already supported by upstream Kubernetes. For some applications the Huge Pages should be allocated to account for considerations of the underlying HW topology. This newer feature is missing from Kubernetes, therefore a gap has been identify which has been added to [Chapter 8.2.8](./chapter08.md/#8.2.8)

Reviewer: ```suggestion
The Reference Model requires the support of Huge Pages in `i.cap.018` which is already supported by upstream Kubernetes. For some applications, Huge Pages should be allocated to account for consideration of the underlying HW topology. This newer feature is missing from Kubernetes, therefore a gap has been identified and added to [Chapter 8.2.8](./chapter08.md/#8.2.8)
```

The Container Runtime is the component that runs within a Host Operating System (OS) and manages the underlying OS functionality, such as cgroups and namespaces (in Linux), in order to provide a service within which container images can be executed and make use of the infrastructure resources (compute, storage, networking and other I/O devices) abstracted by the Host OS, based on API instructions from the kubelet.

> There are a number of different container runtimes. The simplest form, low-level container runtimes, just manage the OS capabilities such as cgroups and namespaces, and then run commands from within those cgroups and namesapces. An example of this type of runtime is runc, which underpins many of the higher-level runtimes and is considered a reference implementation of the [Open Container Initiative (OCI) runtime spec](https://github.com/opencontainers/runtime-spec). This specification includes detail on how an implementation (i.e. an actual container runtime such as runc) must, for example, configure resource shares and limits (e.g. CPU, Memory, IOPS) for the containers that Kubernetes (via the kubelet) schedules on that host. This is important in ensuring the features and capabilities described in [chapter 5 of the RM](../../../ref_model/chapters/chapter05.md) are supported by this RA and delivered by any downstream Reference Implementations (RIs) to the instance types defined in the RM.

Reviewer: ```suggestion
There are a number of different container runtimes. The simplest form, low-level container runtimes, just manage the OS capabilities such as cgroups and namespaces, and then run commands from within those cgroups and namesapces. An example of this type of runtime is runc, which underpins many of the higher-level runtimes and is considered a reference implementation of the [Open Container Initiative (OCI) runtime spec](https://github.com/opencontainers/runtime-spec). This specification includes details on how an implementation (i.e. an actual container runtime such as runc) must, for example, configure resource shares and limits (e.g. CPU, Memory, IOPS) for the containers that Kubernetes (via the kubelet) schedules on that host. This is important to ensure that the features and capabilities described in [chapter 5 of the RM](../../../ref_model/chapters/chapter05.md) are supported by this RA and delivered by any downstream Reference Implementations (RIs) to the instance types defined in the RM.
```

> Where low-level runtimes are used for the execution of a container within an OS, the more complex/complete high-level container runtimes are used for the general management of container images - moving them to where they need to be executed, unpacking them, and then passing them to the low-level runtime, which then executes the container. These high-level runtimes also include a comprehensive API that other applications (e.g. Kubernetes) can use to interact and manage the containers. An example of this type of runtime is containerd, which provides the features described above, before passing off the unpacked container image to runc for execution.

Reviewer: @bfcohen w.r.t. the following sentence, "An example of this type of runtime is containerd, which provides the features described above, before passing off the unpacked container image to runc for execution."
1. there is something missing because the following is obviously incorrect, "An example of this type of runtime is containerd before passing off the unpacked container image to runc for execution."
2. "passing off the unpacked container image to runc for execution" is also described above.

Reviewer 2: Not sure how to solve this.

<a name="3.2.2"></a>
### 3.2.2 Container Networking Services

As all production networking solutions for Kubernetes are based on CNI plugins and the implicit requirement in [4.2.2 Virtual Network Interface Specifications](../../../ref_model/chapters/chapter04.md#422-virtual-network-interface-specifications) that documents the requirement to have the capability to attach several network interfaces to the pods, the CNTT architecture must support a CNI metaplugin/CNI multiplexer.

A CNI metaplugin/CNI multiplexer has the capability to attach several interfaces, using different other CNI plugins, to a pod. Note that the different network characteristics of the interfaces might require different networking technologies, which would potentially require different CNI plugins.

To comply with `req.inf.ntw.08`, inter node communication must be served by a CNI plugin which complies with the default K8s networking assumptions.

> There are two types of low latency and high throughput networks required by telco workloads: Signalling traffic workloads and user plane traffic workloads. Networks used for signalling traffic are more demanding than what a standard overlay network can handle, but still do not need the use of user space networking. Due to the nature of the signalling protocols used, this type of networks require the NAT-less communication documented in `infra.net.cfg.003`. Due to the combination of these two requirements networks with these characteristics must be served by a CNI plugin with IPVLAN or MACVLAN support. On the other hand, the low latency, high throughput networks used for handling the user plane traffic require the capability to use a user space networking technology.

Reviewer: ```suggestion
There are two types of low latency and high throughput networks required by telco workloads: Signalling traffic workloads and user plane traffic workloads. Networks used for signalling traffic are more demanding than what a standard overlay network can handle, but still do not need the use of user space networking. Due to the nature of the signalling protocols used, these type of networks require NAT-less communication documented in `infra.net.cfg.003`. Due to the combination of these two requirements, networks with these characteristics must be served by a CNI plugin with IPVLAN or MACVLAN support. On the other hand, the low latency, high throughput networks used for handling the user plane traffic require the capability to use a user space networking technology.
```

Reviewer: @bfcohen @tomkivlin "Due to the combination of these two requirements, ..." which two requirements?

Reviewer 3: Good question - let me check the commits on that page, it may have been there were two requirements listed previously.

Reviewer 3: The original wording was:
> There are two types of low latency and high throughput networks required by `req.inf.ntw.04`. Network used for signalling traffic <_removed_> require the NAT-less communication stated by `req.inf.ntw.03`. Due to the combination of these two requirements networks with these characteristics must be served by a CNI plugin with IPVLAN or MACVLAN support.

So the two requirements were listed.  

I would suggest modifying the line to:
> There are two types of low latency and high throughput networks required by telco workloads: signalling traffic workloads and user plane traffic workloads. Networks used for signalling traffic are more demanding than what a standard overlay network can handle, but still do not need the use of user space networking. Due to the nature of the signalling protocols used, these type of networks require NAT-less communication documented in `infra.net.cfg.003` and will need to be served by a CNI plugin with IPVLAN or MACVLAN support. On the other hand, the low latency, high throughput networks used for handling the user plane traffic require the capability to use a user space networking technology.

Reviewer 3: Original wording, for completeness: https://github.com/cntt-n/CNTT/commit/686ebb92c37272033b616cc1c00675986cb62c10#diff-44f0d56ff3bb53bd7d3d1facd9e902ffL115

> Note: An SDN solution can manage the pod networks via the Kubernetes API or can have a CNI integration.

The container based architecture must support telecom equipment networking where the CNF networks are set up by the operator's network administrators. This is why, as `req.gen.cnt.05` requires, the architecture must provide a set of abstract management APIs to manage the network connectivity of the CNF pods themselves.

> The API must support multiple tenants and must require elevated acces rights to manipulate infrastructure related API objects as these operations generally require reconfiguration of the physical network infrastructure.

Reviewer: ```suggestion
The API must support multiple tenants and must require elevated access rights to manipulate infrastructure related API objects as these operations generally require reconfiguration of the physical network infrastructure.
```

For those scenarios where data persistence is required, Persistent Volumes (PV) are used in Kubernetes. PVs are resources in a Kubernetes Cluster that are consumed by Persistent Volume Claims (PVCs) and have a lifecycle that is independent of any Pod that uses the PV. A Pod will use a PVC as the volume in the Pod spec; a PVC is a request for persistent storage (a PV) by a Pod. By default, PVs and PVCs are manually created and deleted.

Kubernetes also provides an object called Storage Class, which is created by cluster administrators and maps to storage attributes such as quality-of-service, encryption, data resilience, etc. Storage Classes also enable the dynamic provisioning of Persistent Volumes (as opposed to the default manual creation). This can be beneficial for organisations where the administration of storage is performed separately from the administration of Kubernetes-based workloads.

> There are no restrictions or constraints that Kubernetes places on the storage that can be consumed by a workload, in terms of the requirements that are defined in RM sections [5.2.2](../../../ref_model/chapters/chapter05.md#522-virtual-storage) (software) and [5.4.2](../../../ref_model/chapters/chapter05.md#542-storage-configurations) (hardware). The only point of difference is that Kubernetes does not have a native object storage offering, and and addresing this this capability gap directly is outside of the scope of this RA.

Reviewer: ```suggestion
There are no restrictions or constraints that Kubernetes places on the storage that can be consumed by a workload, in terms of the requirements that are defined in RM sections [5.2.2](../../../ref_model/chapters/chapter05.md#522-virtual-storage) (software) and [5.4.2](../../../ref_model/chapters/chapter05.md#542-storage-configurations) (hardware). The only point of difference is that Kubernetes does not have a native object storage offering, and addressing this capability gap directly is outside of the scope of this RA.
```

<a name="3.2.4"></a>
### 3.2.4 Kubernetes Application package manager

> To manage complex applications consisting of several Pods the reference architecture must provide support for a Kubernetes Application  package manager. The Package Manager would be able to manage the lifecycle of a set of Pods and provide a framework to customise a set of parameters for their deployment. The requirement for this Reference Architecture is to provide a Kubernetes API in conformance with the CNCF Conformance test for the package managers to use in the lifecycle management of the applications they manage. The Reference Architecture does not recommend the usage of a Kubernetes Application package manager with a server side component installed to the Kubernetes cluster (e.g.: Tiller).

Reviewer: ```suggestion
To manage complex applications consisting of several Pods the Reference Architecture must provide support for a Kubernetes Application  package manager. The Package Manager would be able to manage the lifecycle of a set of Pods and provide a framework to customise a set of parameters for their deployment. The requirement for this Reference Architecture is to provide a Kubernetes API that complies with the CNCF Conformance test for the package managers to use in the lifecycle management of the applications they manage. The Reference Architecture does not recommend the usage of a Kubernetes Application package manager with a server side component installed to the Kubernetes cluster (e.g.: Tiller).
```
